R/read_csv.R

```diff
@@ -0,0 +1,262 @@
+#' diagonal of the inverse mass matrix
+#'
+read_sample_info_csv <- function(csv_file) {
+  checkmate::assert_file_exists(csv_file, access = "r", extension = "csv")
+  param_names_read <- FALSE
+  sampling_params_read <- FALSE
+  diagonal_matrix_next <- FALSE
+  diagonal_matrix_read <- FALSE
+  arg_prefix <- ""
+  csv_file_info = list()
+  con  <- file(csv_file, open = "r")
+  while (length(line <- readLines(con, n = 1, warn = FALSE)) > 0) {
+    if (!startsWith(line, "#")) {
+      if(!param_names_read) {
+        param_names_read <- TRUE
+        all_names <- strsplit(line, ",")[[1]]
+        csv_file_info[["sampler_params"]] <- c()
+        csv_file_info[["model_params"]] <- c()
+        for(x in all_names) {
+          if(endsWith(x, "__") && x != "lp__"){
+            csv_file_info[["sampler_params"]] <- c(csv_file_info[["sampler_params"]], x)
+          } else {
+            csv_file_info[["model_params"]] <- c(csv_file_info[["model_params"]], x)
+          }
+        }
+        next;
+      } else {
+        if(diagonal_matrix_read){
+          break;
+        } else {
+          next;
+        }
+      }
+    }
+    # no more sampling settings follow the list of parameters
+    if(!param_names_read) {
+      tmp <- gsub("#", "", line, fixed = TRUE)
+      tmp <- gsub("(Default)", "", tmp, fixed = TRUE)
+      key_val <- grep("=", tmp, fixed = TRUE, value = TRUE)
+      if(length(key_val) == 0) {
+        arg_name <- trimws(tmp)
+        if(arg_prefix != "" && !sampling_params_read) {
+          arg_prefix <- paste(arg_prefix, arg_name, sep = "_")
+        } else {
+          arg_prefix <- trimws(tmp)
+        }
+      } else {
+        key_val <- strsplit(key_val, split = "=", fixed = TRUE)
+        key_val <- rapply(key_val, trimws)
+        if(length(key_val) == 2) {
+          numeric_val <- suppressWarnings(as.numeric(key_val[2]))
+          # marks the end of sampling args
+          if(key_val[1] == "id" || key_val[1] == "init") {
+            arg_prefix <- ""
+            sampling_params_read <- TRUE
+          }
+          if(arg_prefix != ""){
+            setting_name <- paste(arg_prefix, key_val[1], sep = "_")
+          } else {
+            setting_name <- key_val[1]
+          }
+          if(!is.na(numeric_val)) {
+            csv_file_info[[setting_name]] <- numeric_val
+          } else {
+            csv_file_info[[setting_name]] <- key_val[2]
+          }
+        }
+      }
+    }
+    if (regexpr("# Step size = ", line, perl = TRUE) > 0) {
+      key_val <- grep("=", tmp, fixed = TRUE, value = TRUE)
+      csv_file_info[["step_size"]] <- as.numeric(key_val[2])
+    }
+    if (regexpr("# Diagonal elements of inverse mass matrix:", line, perl = TRUE) > 0) {
+      diagonal_matrix_next <- TRUE
+    } else if(diagonal_matrix_next) {
+      csv_file_info$inverse_mass_matrix_diag <- rapply(strsplit(gsub("# ", "", line), ","), as.numeric)
+      diagonal_matrix_next <- FALSE
+      diagonal_matrix_read <- TRUE
+    }
+  }
```

Comment:
What happens if `metric=dense_e` or the user supplied a dense inv_metric?

Reply:
Did not think of that yeah. Yeah a list of the inv mass matrix would be the best way to do this.